dbms/src/Flash/Coprocessor/DAGStorageInterpreter.cpp

```diff
@@ -156,35 +181,27 @@ void DAGStorageInterpreter::execute(DAGPipeline & pipeline)
     else
         learner_read_snapshot = doCopLearnerRead();
 
-    storages_with_structure_lock = getAndLockStorages(settings.schema_version);
-    assert(storages_with_structure_lock.find(logical_table_id) != storages_with_structure_lock.end());
-    storage_for_logical_table = storages_with_structure_lock[logical_table_id].storage;
-
-    std::tie(required_columns, source_columns, is_need_add_cast_column) = getColumnsForTableScan(settings.max_columns_to_read);
-
     analyzer = std::make_unique<DAGExpressionAnalyzer>(std::move(source_columns), context);
 
     FAIL_POINT_PAUSE(FailPoints::pause_after_learner_read);
 
     if (!mvcc_query_info->regions_query_info.empty())
         doLocalRead(pipeline, settings.max_block_size);
 
-    null_stream_if_empty = std::make_shared<NullBlockInputStream>(storage_for_logical_table->getSampleBlockForColumns(required_columns));
+    null_stream_if_empty = std::make_shared<NullBlockInputStream>(storage_table.getLogicalTableStorage()->getSampleBlockForColumns(required_columns));
```

Comment:
```suggestion
    null_stream_if_empty = std::make_shared<NullBlockInputStream>(storage_table.getSampleBlockForColumns(required_columns));
```
Maybe adding `getSampleBlockForColumns` in stroage_table and making "logical table storage" hind from that class?

Reply:
done